### PR TITLE
Quantize Gemma4 per-layer embedding tables for WebGPU (INT4/INT8 configurable)

### PR DESCRIPTION
**Pull request by feich-ms**
## Summary

- Quantize the 35 per-layer embedding tables (`embed_tokens_per_layer_split`) from FP16 to INT4 using `GatherBlockQuantized`, reducing decoder size from 5.8 GB to 2.7 GB for WebGPU
- Add `embedding_bits` parameter to `mobius.build()` so callers can choose INT4 (default) or INT8 without editing source code
- Add `quantize_embedding_rtn()` utility for block-wise RTN quantization of embedding tables (supports 4-bit and 8-bit, symmetric and asymmetric)
- Add `QuantizedScaledWordEmbedding` class that wraps `GatherBlockQuantized` with the Gemma4 embed scaling

## Details

The Gemma4 E2B multimodal model splits a fused [262144, 35×256] embedding table into 35 separate [262144, 256] tables for WebGPU's 256 MiB buffer limit. At FP16, these tables alone take 4.48 GB (77% of the decoder). This PR quantizes them to INT4 (560 MB) or INT8 (1.12 GB) using the `GatherBlockQuantized` custom op.

The `embedding_bits` parameter can be set via:
- `mobius.build(..., embedding_bits=4)` (Python API)
- `"embedding_bits": 4` in the MobiusBuilder Olive pass config JSON

When omitted, defaults to INT4 for backward compatibility.

## Test plan

- [x] Build INT4 model → decoder 2.7 GB, generates clean 890+ tokens
- [x] Build INT8 model → decoder 3.8 GB, generates clean 1024+ tokens
- [x] Omitting `embedding_bits` preserves existing INT4 default behavior
- [ ] CI passes (no regressions to non-WebGPU paths)

### CODE REVIEW

File: src/mobius/_builder.py [hunk 1]
```diff
@@ -354,6 +354,7 @@ def build(
     execution_provider: str = "default",
     trace_optimization: bool = False,
     text_only: bool = False,
+    embedding_bits: int | None = None,
 ) -> ModelPackage:
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/_builder.py:359]: Fixed — added full docstring entry for embedding_bits in the Args section.

File: src/mobius/_builder.py [hunk 2]
```diff
@@ -576,6 +577,23 @@ def build(
...
+                quantization=QuantizationConfig(
+                    bits=embedding_bits, group_size=32, quant_method="mobius",
+                    sym=False, quantize_embeddings=True,
+                ),
+            )
+        else:
+            qc = dataclasses.replace(
+                config.quantization, bits=embedding_bits, quantize_embeddings=True,
+            )
+            config = dataclasses.replace(config, quantization=qc)
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/_builder.py]: Good catch. Verified that quant_method="mobius" is not recognized by any code that swaps Linear to QuantizedLinear (only gptq/awq trigger that). The quantize_embeddings flag only affects Gather→GatherBlockQuantized in models that explicitly check it. This path only fires when embedding_bits is explicitly passed, so it's opt-in. Added validation that embedding_bits must be 4 or 8.

File: src/mobius/tasks/_gemma4.py [hunk 3]
```diff
@@ -222,6 +222,52 @@ def build(
...
+                pass
+            elif qc is not None:
+                # Existing quantization config without embedding quantization —
+                # enable it, defaulting to INT4
+                config.quantization.quantize_embeddings = True
+            else:
+                config.quantization = QuantizationConfig(
+                    bits=4, group_size=32, quant_method="mobius", sym=False,
+                    quantize_embeddings=True,
+                )
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/tasks/_gemma4.py]: Fixed — the elif branch now explicitly sets bits=4, group_size=32, sym=False so behavior is deterministic regardless of what the existing config had.

File: src/mobius/models/gemma4.py [hunk 4]
```diff
@@ -2015,8 +2054,25 @@ def preprocess_weights(
...
+                if quantize_per_layer:
+                    from mobius._weight_utils import quantize_embedding_rtn
+
+                    for i, chunk in enumerate(chunks):
+                        qweight, scales, zero_points = quantize_embedding_rtn(
+                            chunk.contiguous(),
+                            bits=qc.bits,
+                            block_size=qc.group_size,
+                            symmetric=qc.sym,
+                        )
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/models/gemma4.py:2432]: Fixed — added bits validation (must be 4 or 8) before calling quantize_embedding_rtn. The utility now also calls .detach().cpu() defensively for non-CPU tensors.

File: src/mobius/models/gemma4.py [hunk 5]
```diff
@@ -2809,8 +2865,25 @@ def preprocess_weights(
...
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/models/gemma4.py:3253]: Fixed — same bits validation added here.

File: src/mobius/_weight_utils.py [hunk 6]
```diff
@@ -658,6 +658,108 @@ def preprocess_gptq_weights(
...
+    assert bits in (4, 8), f"bits must be 4 or 8, got {bits}"
+    assert weight.ndim == 2
+    num_embeddings, embedding_dim = weight.shape
+    assert embedding_dim % block_size == 0, (
+        f"embedding_dim ({embedding_dim}) must be divisible by block_size ({block_size})"
+    )
+
+    n_blocks = embedding_dim // block_size
+    w = weight.float().numpy()
+
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/_weight_utils.py]: Fixed — replaced all assert with ValueError and added .detach().cpu() before .numpy() to handle non-CPU tensors.

File: src/mobius/_weight_utils.py [hunk 7]
```diff
@@ -658,6 +658,108 @@ def preprocess_gptq_weights(
     return result
 
 
+def quantize_embedding_rtn(
+    weight: torch.Tensor,
+    bits: int = 4,
+    block_size: int = 32,
+    symmetric: bool = False,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor | None]:
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply src/mobius/_weight_utils.py:714]: Added 9 unit tests in _weight_utils_test.py::TestQuantizeEmbeddingRTN covering: INT4/INT8 x symmetric/asymmetric shape checks, dtype validation, error cases (invalid bits, non-2D, indivisible dim), and INT8 roundtrip accuracy.
